File: 5.longest-palindromic-substring.js
/*
 * @lc app=leetcode id=5 lang=javascript
 *
 * [5] Longest Palindromic Substring
 */

// @lc code=start
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function (s) {
  const len = s.length;
  const dp = Array.from({ length: len }, () => []);
  let lps = "";
  for (let j = 0; j < len; ++j) {
    for (let i = 0; i <= j; ++i) {
      //classic  deduce
      if (s[i] === s[j] && (j - i <= 1 || dp[i + 1][j - 1] === true)) {
        dp[i][j] = true;
        const substr = s.slice(i, j + 1);
        if (substr.length > lps.length) {
          lps = substr;
        }
      }
    }
  }
  return lps;
};
// @lc code=end

File: 5.longest-palindromic-substring.rb
#
# @lc app=leetcode id=5 lang=ruby
#
# [5] Longest Palindromic Substring
#

# @lc code=start
# @param {String} s
# @return {String}
def longest_palindrome(s)
  len = s.length
  dp = Array.new(len) { [] }
  lpd = ''
  (0..len - 1).each do |j|
    j.downto(0).each do |i|
      if s[i] == s[j] && (j - i <= 2 || dp[i + 1][j - 1])
        dp[i][j] = true
        lpd = s[i..j] if lpd.length < j - i + 1
      end
    end
  end
  lpd
end
# @lc code=end
